csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/desugar/Foreach.qll

```diff
@@ -0,0 +1,516 @@
+                                               TTranslatedCompilerGeneratedElement {
+  override ForeachStmt generatedBy;
+  
+  TranslatedForeachCurrent() {
+    this = TTranslatedCompilerGeneratedElement(generatedBy, 5)
+  }
+  
+  override Type getCallResultType() { 
+    result = generatedBy.getAVariable().getType()
+  }
+  
+  override TranslatedExpr getArgument(int id) { none() }
+
+  override predicate hasArguments() { none() }
+  
+  override TranslatedExprBlueprint getQualifier() { 
+    exists(TranslatedForeachCurrentEnumAcc acc |
+      acc.getAST() = generatedBy and
+      result = acc
+    )
+  }
+  
+  override Instruction getQualifierResult() {
+    result = getQualifier().getResult()
+  }
+  
+  override Callable getInstructionFunction(InstructionTag tag) {
+    tag = CallTargetTag() and
+    exists(Property prop | 
+      prop.getName() = "Current" and 
+      result = prop.getGetter()
+    )
+  }
+}
+
+/**
+ * The translation of the call to dispose (inside the finally block)
+ */
+private class TranslatedForeachDispose extends TranslatedCompilerGeneratedCall, 
+                                               TTranslatedCompilerGeneratedElement {
+  override ForeachStmt generatedBy;
+  
+  TranslatedForeachDispose() {
+    this = TTranslatedCompilerGeneratedElement(generatedBy, 6)
+  }
+  
+  override Callable getInstructionFunction(InstructionTag tag) {
+    tag = CallTargetTag() and
+    exists(Callable dispose | 
+      dispose.getName() = "Dispose" and 
+      result = dispose
+    )
+  }
+  
+  final override Type getCallResultType() { 
+    result instanceof VoidType
+  }
+  
+  override TranslatedExpr getArgument(int id) { none() }
+  
+  override predicate hasArguments() { none() }
+  
+  override TranslatedExprBlueprint getQualifier() { 
+    exists(TranslatedForeachDisposeEnumAcc acc |
+      acc.getAST() = generatedBy and
+      result = acc
+    )
+  }
+  
+  override Instruction getQualifierResult() {
+    result = getQualifier().getResult()
+  }
+}
+
+/**
+ * The condition for the while, ie. a call to MoveNext.
+ */
+private class TranslatedForeachWhileCondition extends TranslatedCompilerGeneratedValueCondition, 
+                                                      TTranslatedCompilerGeneratedElement {
+  override ForeachStmt generatedBy;
+  
+  TranslatedForeachWhileCondition() {
+    this = TTranslatedCompilerGeneratedElement(generatedBy, 7)
+  }
+  
+  override TranslatedCompilerGeneratedCall getValueExpr() {
+    exists(TranslatedForeachMoveNext mn |
+      mn.getAST() = generatedBy and
+      result = mn
+    )
+  }
+  
+  override Instruction valueExprResult() {
+    result = getValueExpr().getResult()
+  }
+}
+
+/**
+ * Class that represents that translation of the declaration that happens before the `try ... finally` block (the
+ * declaration of the `temporary` enumerator variable)
+ */
+private class TranslatedForeachEnum extends TranslatedCompilerGeneratedDeclaration, 
```

Comment:
I find the name misleading (thinking of `enum`s); `TranslatedForeachEnumerator` would be better.

Reply:
Indeed, fixed.